trino/client.py

```diff
@@ -864,19 +864,22 @@ def _double_map_func(self):
     def _timestamp_map_func(self, column, col_type):
         datetime_default_size = 20  # size of 'YYYY-MM-DD HH:MM:SS.' (the datetime string up to the milliseconds)
         pattern = "%Y-%m-%d %H:%M:%S"
-        ms_size, ms_to_trim = self._get_number_of_digits(column)
+        ms_size, ms_to_trim, ms_div = self._get_number_of_digits(column)
         if ms_size > 0:
             pattern += ".%f"
 
-        dt_size = datetime_default_size + ms_size - ms_to_trim
+        dt_size = datetime_default_size
         dt_tz_offset = datetime_default_size + ms_size
         if 'with time zone' in col_type:
-
             if ms_to_trim > 0:
                 return lambda val: \
-                    [datetime.strptime(val[:dt_size] + val[dt_tz_offset:], pattern + ' %z')
+                    [datetime.strptime(val[:dt_size]
+                     + str(round(int(val[dt_size:dt_tz_offset]) / ms_div))
```

Comment:
note: `round` returns floats so it'll lose precision and cause incorrect rounding in some cases.

Reply:
The round returns a float but the code only cares about the integer part. It takes the total number of digits after the decimal (e.g. `123456789`), divides it by `ms_div` to only keep 6 digits (e.g. `123456.789`), rounds this (`123457`) and converts it to a string. So to lose precision you would need to have a timestamp way more precise than 12 as `round(0.499999999999999999)` returns 0 and `round(0.5000000000000001)` return 1.

Reply:
That's not the issue I'm thinking of. Try `round(2.675, 2)`. It can happen for any other value since floats cannot represent all values exactly.